cobbler/actions/buildiso/__init__.py

```diff
@@ -0,0 +1,242 @@
+        ISO.
+        """
+        self.logger.info("copying syslinux files")
+
+        files_to_copy = ["isolinux.bin", "menu.c32", "chain.c32", "ldlinux.c32", "libcom32.c32", "libutil.c32"]
+        optional_files = ["ldlinux.c32", "libcom32.c32", "libutil.c32"]
+        syslinux_folders = [pathlib.Path(self.api.settings().syslinux_dir),
+                            pathlib.Path(self.api.settings().syslinux_dir).joinpath("modules/bios/"),
+                            pathlib.Path("/usr/lib/syslinux/"),
+                            pathlib.Path("/usr/lib/ISOLINUX/")]
+
+        # file_copy_success will be used to check for missing files
+        file_copy_success: Dict[str, bool] = {f: False for f in files_to_copy if f not in optional_files}
+        for syslinux_folder in syslinux_folders:
+            if syslinux_folder.exists():
+                for file_to_copy in files_to_copy:
+                    source_file = syslinux_folder.joinpath(file_to_copy)
+                    if source_file.exists():
+                        utils.copyfile(str(source_file), os.path.join(self.isolinuxdir, file_to_copy))
+                        file_copy_success[file_to_copy] = True
+
+        unsuccessful_copied_files = [k for k, v in file_copy_success.items() if not v]
+        if len(unsuccessful_copied_files) > 0:
+            self.logger.error("The following files were not found: \"%s\"", "\", \"".join(unsuccessful_copied_files))
+            raise FileNotFoundError("Required file(s) not found. Please check your syslinux installation")
+
+        self.logger.info("copying GRUB2 files")
+        bootloader_directory = pathlib.Path(self.api.settings().bootloaders_dir)
+        grub_efi = bootloader_directory.joinpath("grub", "grub.efi")
+        buildiso_directory = pathlib.Path(self.api.settings().buildisodir)
+        grub_target = buildiso_directory.joinpath("grub", "grub.efi")
+        if grub_efi.exists():
+            utils.copyfile(str(grub_efi), str(grub_target), symlink=True)
+        else:
+            self.logger.error("The following files were not found: \"%s\"", grub_efi)
+            raise FileNotFoundError("Required file(s) not found. Please check your GRUB 2 installation")
+
+    def __prepare_buildisodir(self, buildisodir: str) -> str:
+        """
+        This validated the path and type of the buildiso directory and then (re-)creates the apropiate directories.
+
+        :param buildisodir: The directory in which the build of the ISO takes place.
+        :raises ValueError: In case the specified directory does not exist.
+        :raises TypeError: In case the specified argument is not of type str.
+        :return: The validated and normalized directory with appropriate subfolders provisioned.
+        """
+        if not isinstance(buildisodir, str):
+            raise TypeError("buildisodir needs to be of type str!")
+        if not buildisodir:
+            buildisodir = self.api.settings().buildisodir
```

Comment:
Well the problem is that we can't give the real default value here. The default value would be in the settings and the settings are not accessible at this point. The reason for this is that the default parameter is set when [the function definition is being executed](https://docs.python.org/3/reference/compound_stmts.html#function-definitions). Thus we need to set this in the method body.

Do you have any proposal how to fix that problem? Because an empty path is invalid and thus I used it as the default case.

Reply:
It's fine to set it inside the function body, I don't find that problematic at all. My comment was more about the how we get to setting a fallback/default value from the settings. Let me try to clarify that a bit:

Based on the function signature, `buildisodir` is a `str`. This conditional statement only evaluates in one case: `bulidisodir == ""`. The fallback value is only set in this one case. An empty `str` might as well be the default for the `buildisodir` in the argument list of `__prepare_buildisodir`, making `buildisodir` an optional parameter to pass.

> The reason for this is that the default parameter is set when the function definition is being executed. Thus we need to set this in the method body.

You're completely right about that! It's an easy mistake to make when using empty lists/dicts etc. as default parameters.

Reply:
Ah you mean it like this. Yeah of course we can do exactly that.

Reply:
Fixed